llvm/include/llvm/CAS/ThreadSafeAllocator.h

```diff
@@ -23,7 +23,8 @@ namespace cas {
 /// added thread-safety to the allocator itself. Probably best/reasonable to use
 /// a fair lock (std::mutex) when allocating a new slab, but otherwise use
 /// atomics to make Allocate() lock-free.
-template <class AllocatorType> class ThreadSafeAllocator {
+template <class AllocatorType>
+class ThreadSafeAllocator : public AllocatorType {
```

Comment:
I feel like using `public` inheritance could be confusing. Any APIs that come through via inheritance will not be thread-safe.

Is there a specific reason you added that? Can you provide it without losing thread-safety?

Reply:
Another option would be to add the following as a catch-all for other APIs:
```
void ThreadSafeAllocator<AllocatorT>::applyLocked(llvm::function_ref<void(AllocatorT &Alloc)> Fn) {
  std::unique_lock<std::mutex> LockGuard(Lock);
  Fn(Alloc);
}
```

Reply:
Yup, that concern is in the commit message. Just want some interface from Allocator for testing purpose. Otherwise the test might be a bit too simple.

Reply:
I think adding `applyLocked()` should work well then; you could even change the `Allocate` APIs to be implemented using `applyLocked()` (depending on whether that made them look simpler).

Reply:
I can't quite understand the `appleLocked` idea as currently there is no way to access the underlying allocator from ThreadSafeAllocator. And also maybe we don't need to lock a `const` method.

Maybe it would be good to understand what your concern was for the FIXME at https://github.com/apple/llvm-project/blob/5cf200ead177a63d85b5cbbdbf24d280347c87c3/llvm/include/llvm/CAS/ThreadSafeAllocator.h#L21, and we can just test the improved version instead.

Reply:
> I can't quite understand the `appleLocked` idea as currently there is no way to access the underlying allocator from ThreadSafeAllocator.

The idea is to add `applyLocked()` in order to give access to the underlying allocator. See `SharedStream` in clang/tools/clang-scan-deps/ClangScanDeps.cpp, which has been copy-pasted into a couple of places in this branch, for another example.

Basically you just add this method:
```
template <class AllocT> class ThreadSafeAllocator {
public:
  void applyLocked(llvm::function_ref<void(AllocatorT &Alloc)> Fn) {
    LockGuard Lock(Flag);
    Fn(Alloc);
  }
// rest of implementation
};
```
then from your test you can do:
```
  ThreadSafeAlloctor<BumpPtrAllocator> Alloc;
  // do allocations

  // Check the number of allocations.
  Alloc.applyLocked([](BumpPtrAllocator &BumpAlloc) {
    EXPECT_EQ(4950U * sizeof(int), BumpAlloc.getBytesAllocated());
  });
```

> And also maybe we don't need to lock a `const` method.

You need to lock anything that you want to be thread-safe, given that the underlying allocator is not thread-safe. Given that "thread-safe" is in the name of this type, probably every action should be made thread-safe.

> Maybe it would be good to understand what your concern was for the FIXME at
> 
> https://github.com/apple/llvm-project/blob/5cf200ead177a63d85b5cbbdbf24d280347c87c3/llvm/include/llvm/CAS/ThreadSafeAllocator.h#L21
> 
> , and we can just test the improved version instead.

Probably that should be a TODO instead.

The "concern" is that:
- Locking is expensive.
- Hand-implementing an unfair lock via `atomic_flag` (what the implementation apparently does!) is also expensive. If the lock actually has contention, then it's potentially very expensive.
- It'd be better to implement an algorithm for a bump-ptr allocator that was lock-free in the common case (bumping an existing allocation). It'd be fine to take a lock when adding a new slab.

Here's a sketch that would possibly work:
```
class ThreadSafeBumpPtrAllocator {
public:
  void *Allocate(size_t N = 1) {
    // For big "N", just use a separate slab. Whatever the math is in BumpPtrAllocator is
    // probably fine. 
    if (N > SlabSize / 2) {
         std::scoped_lock<std::mutex> Lock(Mutex);
         Allocations.push_back(malloc(N));
         return Allocations.back().get();
    }

    uintptr_t SlabBegin = CurrentBegin;
    uintptr_t SlabEnd = SlabBegin + SlabSize;
    uintptr_t Allocation = Current.fetch_add(N);
    if (Allocation >= SlabBegin && Allocation < SlabEnd) // Check for race
      if (Allocation + N < SlabEnd) // Check there was space in the slab.
        return (void*)Allocation;

    // Either raced with allocating a new slab, or we need a new slab.
    std::scoped_lock<std::mutex> Lock(Mutex);
    SlabBegin = CurrentBegin;
    SlabEnd = SlabBegin + SlabSize;

    // Try again to allocate with the existing slab in case it failed due to a race.
    if (!(Allocation >= SlabBegin && Allocation < SlabEnd))
      Allocation = Current.fetch_add(N); // Fetch again only if the slab changed.
    if (Allocation + N < SlabEnd) // Check for space in the slab.
      return (void *)Allocation;

    // Create a new slab.
    Allocations.push_back(malloc(SlabSize));
    SlabBegin = (uintptr_t)Allocations.back();
    Current = Allocation = SlabBegin + N;
    CurrentBegin = SlabBegin;
    return (void*)Allocation;
  }

private:
  const size_t SlabSize; // Never changed.
  std::atomic<uintptr_t> Current; // Changed with or without lock.
  std::atomic<uintptr_t> CurrentBegin; // Always locked for changes.

  std::mutex Mutex;
  std::vector<std::unique_ptr<char[]>> Allocations;
};
```

Reply:
Nice. I wasn't aware that pattern. Updated PR to use that.

Also clarify the FIXME  -> TODO so we can revisit if we found there are more contentions than expected in allocator.